exchanges/okx/okx.go

```diff
@@ -0,0 +1,4453 @@
+func (ok *Okx) PlaceStopOrder(ctx context.Context, arg *AlgoOrderParams) (*AlgoOrder, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.OrderType != "conditional" {
+		return nil, errInvalidOrderType
+	}
+	if arg.TakeProfitTriggerPrice == 0 {
+		return nil, errMissingTakeProfitTriggerPrice
+	}
+	if arg.TakeProfitTriggerPriceType == "" {
+		return nil, errMissingTakeProfitOrderPrice
+	}
+	return ok.PlaceAlgoOrder(ctx, arg)
+}
+
+// PlaceTrailingStopOrder to place trailing stop order
+func (ok *Okx) PlaceTrailingStopOrder(ctx context.Context, arg *AlgoOrderParams) (*AlgoOrder, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.OrderType != "move_order_stop" {
+		return nil, errInvalidOrderType
+	}
+	if arg.CallbackRatio == 0 &&
+		arg.CallbackSpreadVariance == "" {
+		return nil, errors.New("either \"callbackRatio\" or \"callbackSpread\" is allowed to be passed")
+	}
+	return ok.PlaceAlgoOrder(ctx, arg)
+}
+
+// PlaceIcebergOrder to place iceburg algo order
+func (ok *Okx) PlaceIcebergOrder(ctx context.Context, arg *AlgoOrderParams) (*AlgoOrder, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.OrderType != "iceberg" {
+		return nil, errInvalidOrderType
+	}
+	if arg.PriceVariance == "" && arg.PriceSpread == "" {
+		return nil, errEitherPriceVarianceOrPriceSpreadRequired
+	}
+	if arg.SizeLimit <= 0 {
+		return nil, errMissingSizeLimit
+	}
+	if arg.PriceLimit <= 0 {
+		return nil, errInvalidPriceLimit
+	}
+	return ok.PlaceAlgoOrder(ctx, arg)
+}
+
+// PlaceTWAPOrder to place TWAP algo orders
+func (ok *Okx) PlaceTWAPOrder(ctx context.Context, arg *AlgoOrderParams) (*AlgoOrder, error) {
+	if arg.PriceVariance == "" && arg.PriceSpread == "" {
+		return nil, errEitherPriceVarianceOrPriceSpreadRequired
+	}
+	if arg.OrderType != "twap" {
+		return nil, errInvalidOrderType
+	}
+	if arg.SizeLimit <= 0 {
+		return nil, errors.New("missing required parameter \"size limit\"")
+	}
+	if arg.PriceLimit <= 0 {
+		return nil, errInvalidPriceLimit
+	}
+	if ok.GetIntervalEnum(arg.TimeInterval) == "" {
+		return nil, errMissingIntervalValue
+	}
+	return ok.PlaceAlgoOrder(ctx, arg)
+}
+
+// TriggerAlgoOrder fetches algo trigger orders for SWAP market types.
+func (ok *Okx) TriggerAlgoOrder(ctx context.Context, arg *AlgoOrderParams) (*AlgoOrder, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.OrderType != "trigger" {
+		return nil, errInvalidOrderType
+	}
+	if arg.TriggerPrice <= 0 {
+		return nil, errInvalidTriggerPrice
+	}
+	if arg.TriggerPriceType != "last" &&
+		arg.TriggerPriceType != "index" &&
+		arg.TriggerPriceType != "mark" {
+		arg.TriggerPriceType = ""
+	}
+	return ok.PlaceAlgoOrder(ctx, arg)
+}
+
+// CancelAdvanceAlgoOrder Cancel unfilled algo orders
+// A maximum of 10 orders can be canceled at a time.
+// Request parameters should be passed in the form of an array.
+func (ok *Okx) CancelAdvanceAlgoOrder(ctx context.Context, args []AlgoOrderCancelParams) ([]AlgoOrder, error) {
```

Comment:
You should consider passing in the rate limit as well, `cancelAdvanceAlgoOrderEPL` is not used as a result of this